crates/oxc_linter/src/rules/promise/no_promise_in_callback.rs

```diff
@@ -0,0 +1,168 @@
+use oxc_ast::ast::FormalParameters;
+use oxc_ast::AstKind;
+use oxc_diagnostics::OxcDiagnostic;
+use oxc_macros::declare_oxc_lint;
+use oxc_span::{GetSpan, Span};
+
+use crate::utils::is_promise;
+use crate::{context::LintContext, rule::Rule, AstNode};
+
+fn no_promise_in_callback_diagnostic(span: Span) -> OxcDiagnostic {
+    OxcDiagnostic::warn("Avoid using promises inside of callbacks.")
+        .with_help("Use either promises or callbacks exclusively for handling asynchronous code.")
+        .with_label(span)
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct NoPromiseInCallback;
+
+declare_oxc_lint!(
+    /// ### What it does
+    /// Disallows the use of Promises within error-first callback functions.
+    ///
+    /// ### Why is this bad?
+    /// Mixing Promises and callbacks can lead to unclear and inconsistent code.
+    /// Promises and callbacks are different patterns for handling asynchronous code.
+    /// Mixing them makes the logic flow harder to follow and complicates error handling,
+    /// as callbacks rely on an error-first pattern, while Promises use `catch`.
+    ///
+    /// ### Examples
+    ///
+    /// Examples of **incorrect** code for this rule:
+    /// ```js
+    /// doSomething((err, val) => {
+    ///   if (err) console.error(err)
+    ///   else doSomethingElse(val).then(console.log)
+    /// })
+    /// ```
+    ///
+    /// Examples of **correct** code for this rule:
+    /// ```js
+    /// promisify(doSomething)()
+    ///   .then(doSomethingElse)
+    ///   .then(console.log)
+    ///   .catch(console.error)
+    /// ```
+    NoPromiseInCallback,
+    suspicious,
+);
+
+impl Rule for NoPromiseInCallback {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        let AstKind::CallExpression(call_expr) = node.kind() else {
+            return;
+        };
+
+        if is_promise(call_expr).is_none() {
+            return;
+        }
+
+        // When a Promise is returned in a ReturnStatement, the function is most likely
+        // being used as part of a Promise chain rather than as a callback function.
+        // To avoid false positives, this case is intentionally excluded from the scope of this rule.
+        if let Some(AstKind::ReturnStatement(_)) = ctx.nodes().parent_kind(node.id()) {
+            return;
+        };
+
+        let mut ancestors = ctx.nodes().ancestors(node.id());
+        if ancestors.any(|node| is_callback_function(node, ctx)) {
+            ctx.diagnostic(no_promise_in_callback_diagnostic(call_expr.callee.span()));
+        }
+    }
+}
+
+fn is_callback_function<'a>(node: &AstNode<'a>, ctx: &LintContext<'a>) -> bool {
```

Comment:
In the original ESLint implementation, some functions were located in the `lib` directory.
I moved them into this rule file because they are not used by other rules.